File: src/storeFx/subscribeToAudioEvents.ts
import { deepEqual } from '@tanstack/react-router';

import * as Fx from '../fx/fx';
import type { Store } from '../store/types';
import { goToNextSong } from './goToNextSong';
import { audio } from './playerContext';
import { saveCurrentTimeThrottled } from './saveCurrentTime';

export function subscribeToAudioEvents() {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
    const abortController = new AbortController();

    audio.addEventListener(
      'ended',
      () => {
        Fx.runAsync(goToNextSong(), { store }).then(result => {
          if (result.TAG !== 0) throw new Error();
        });
      },
      { signal: abortController.signal },
    );

    audio.addEventListener(
      'emptied',
      () => {
        store.setState(prevState => ({
          player: {
            ...prevState.player,
            buffered: [],
            currentTime: 0,
            duration: undefined,
            muted: prevState.player.muted,
            paused: true,
            volume: prevState.player.volume,
          },
        }));
      },
      { signal: abortController.signal },
    );

    audio.addEventListener(
      'durationchange',
      () => {
        store.setState(prevState => ({
          player: {
            ...prevState.player,
            duration: Number.isFinite(audio.duration)
              ? audio.duration
              : undefined,
          },
        }));
      },
      { signal: abortController.signal },
    );

    audio.addEventListener(
      'pause',
      () => {
        store.setState(prevState => ({
          player: {
            ...prevState.player,
            paused: true,
          },
        }));
      },
      { signal: abortController.signal },
    );

    audio.addEventListener(
      'play',
      () => {
        store.setState(prevState => ({
          player: {
            ...prevState.player,
            paused: false,
          },
        }));
      },
      { signal: abortController.signal },
    );

    let updateCurrentTimeRaf: number | undefined;
    audio.addEventListener(
      'timeupdate',
      () => {
        if (updateCurrentTimeRaf != null) {
          cancelAnimationFrame(updateCurrentTimeRaf);
        }

        updateCurrentTimeRaf = requestAnimationFrame(() => {
          updateCurrentTimeRaf = undefined;

          store.setState(prevState => ({
            player: {
              ...prevState.player,
              currentTime: audio.currentTime,
            },
          }));
        });

        saveCurrentTimeThrottled(store);
      },
      { signal: abortController.signal },
    );

    store.setState(prevState => ({
      player: {
        ...prevState.player,
        currentTime: audio.currentTime,
        duration: Number.isFinite(audio.duration) ? audio.duration : undefined,
        muted: audio.muted,
        paused: audio.paused,
        volume: audio.volume,
      },
    }));

    const interval = setInterval(() => {
      if (audio.src === '') return;

      const buffered = Array.from(
        { length: audio.buffered.length },
        (_, i) => ({
          end: audio.buffered.end(i),
          start: audio.buffered.start(i),
        }),
      );

      const state = store.getState();

      if (!deepEqual(buffered, state.player.buffered)) {
        store.setState(prevState => ({
          player: {
            ...prevState.player,
            buffered,
          },
        }));
      }
    }, 500);

    return Fx.Ok(() => {
      clearInterval(interval);
      abortController.abort();
    });
  });
}

File: src/storeFx/subscribeToAudioEvents.ml
open Fx
open Monad_syntax

let make () =
  let* Deps.{ store } = ask () in
  let abortController = Fetch.AbortController.make () in
  let signal = abortController |. Fetch.AbortController.signal in
  let listenerOptions = DomExtra.listenerOptions ~signal () in
  let addEventListener listener =
    PlayerContext.audio |. DomExtra.addEventListener listener listenerOptions
  in
  addEventListener
    (`ended
       (fun _ ->
         GoToNextSong.make ()
         |. runAsync ~deps:{ store }
         |> Js.Promise.then_ (fun result ->
             result |> Result.get_ok |> Js.Promise.resolve)
         |> ignore));
  addEventListener
    (`emptied
       (fun _ ->
         store
         |. Zustand.setState (fun prevState ->
             {
               prevState with
               player =
                 {
                   prevState.player with
                   buffered = [||];
                   currentTime = 0.0;
                   duration = None;
                   muted = prevState.player.muted;
                   paused = true;
                   volume = prevState.player.volume;
                 };
             })));
  addEventListener
    (`durationchange
       (fun _ ->
         store
         |. Zustand.setState (fun prevState ->
             {
               prevState with
               player =
                 {
                   prevState.player with
                   duration =
                     (if
                        PlayerContext.audio
                        |. DomExtra.HtmlAudioElement.duration
                        |> Js.Float.isFinite
                      then
                        PlayerContext.audio
                        |. DomExtra.HtmlAudioElement.duration |> Option.some
                      else None);
                 };
             })));
  addEventListener
    (`pause
       (fun _ ->
         store
         |. Zustand.setState (fun prevState ->
             { prevState with player = { prevState.player with paused = true } })));
  addEventListener
    (`play
       (fun _ ->
         store
         |. Zustand.setState (fun prevState ->
             {
               prevState with
               player = { prevState.player with paused = false };
             })));
  let updateCurrentTimeRaf = ref None in
  addEventListener
    (`timeupdate
       (fun _ ->
         !updateCurrentTimeRaf |> Option.iter Webapi.cancelAnimationFrame;
         updateCurrentTimeRaf :=
           Some
             (Webapi.requestCancellableAnimationFrame (fun _ ->
                  updateCurrentTimeRaf := None;
                  store
                  |. Zustand.setState (fun prevState ->
                      {
                        prevState with
                        player =
                          {
                            prevState.player with
                            currentTime =
                              PlayerContext.audio
                              |. DomExtra.HtmlAudioElement.currentTime;
                          };
                      })));
         store |> SaveCurrentTime.make |> ignore));
  store
  |. Zustand.setState (fun prevState ->
      {
        prevState with
        player =
          {
            prevState.player with
            currentTime =
              PlayerContext.audio |. DomExtra.HtmlAudioElement.currentTime;
            paused = PlayerContext.audio |. DomExtra.HtmlAudioElement.paused;
            duration =
              (if
                 Js.Float.isFinite
                   (PlayerContext.audio |. DomExtra.HtmlAudioElement.duration)
               then
                 Some (PlayerContext.audio |. DomExtra.HtmlAudioElement.duration)
               else None);
            muted = PlayerContext.audio |. DomExtra.HtmlAudioElement.muted;
            volume = PlayerContext.audio |. DomExtra.HtmlAudioElement.volume;
          };
      });
  let interval =
    Js.Global.setInterval 500 ~f:(fun () ->
        let src = PlayerContext.audio |. DomExtra.HtmlAudioElement.src in
        if src != "" then
          let buffered =
            PlayerContext.audio |. DomExtra.HtmlAudioElement.buffered
          in
          let buffered =
            buffered |> DomExtra.TimeRanges.length
            |. Belt.Array.init (fun i ->
                Store.State.
                  {
                    end_ = buffered |. DomExtra.TimeRanges.end_ i;
                    start = buffered |. DomExtra.TimeRanges.start i;
                  })
          in
          let Store.State.{ player; _ } = store |. Zustand.getState in
          if not (Router.deepEqual buffered player.buffered) then
            store
            |. Zustand.setState (fun prevState ->
                { prevState with player = { prevState.player with buffered } }))
  in
  Ok
    (fun () ->
      Js.Global.clearInterval interval;
      abortController |> Fetch.AbortController.abort)
